src/os/bluestore/BlueFS.cc

```diff
@@ -2729,14 +2726,18 @@ void BlueFS::_compact_log_async_LD_LNF_D() //also locks FW for new_writer
   // Part 0.
   // Lock the log and forbid its expansion and other compactions
 
+  // lock log's run-time structures for a while
+  log.lock.lock();
+
+  // Extend log in case of having a big transaction waiting before starting compaction.
+  _maybe_extend_log();
+
```

Comment:
if we add this below `bool old_is_comp = std::atomic_exchange(&log_is_compacting, true);` it will cause a infinite loop in `_extend_log`. The return below the atomic_exchange would't unlock the log even without these changes so we may have a bug there.

Reply:
hmm... not that answering this question is important but mostly out of curiosity...

I'm talking about reordering the current code in the following way:

```
  // only one compaction allowed at one time
  bool old_is_comp = std::atomic_exchange(&log_is_compacting, true);
  if (old_is_comp) {
    dout(10) << __func__ << " ongoing" <<dendl;
    log.lock.unlock();
    return;
  }

 // lock log's run-time structures for a while
  log.lock.lock();

  // Extend log in case of having a big transaction waiting before starting compaction.
  _maybe_extend_log();

```
IIUC _maybe_extent_log() doesn't access log_is_compacting flag in any way.  I can see waiting on log_forbidden_to_expand but not on log_is_compacting. So how this rearrangement would cause any infinite loop?
By setting the flag a bit earlier we just indicate to others that compaction is running (or will happen after maybe_extend_log). So they don't need to run it at that moment. Suggested reordering just extends that gap when other compactions are forbidden.